src/main/java/jcifs/smb1/netbios/Lmhosts.java

```diff
@@ -0,0 +1,158 @@
+    private static LogStream log = LogStream.getInstance();
+
+    /**
+     * This is really just for {@link jcifs.smb1.UniAddress}. It does
+     * not throw an {@link java.net.UnknownHostException} because this
+     * is queried frequently and exceptions would be rather costly to
+     * throw on a regular basis here.
+     */
+
+    public synchronized static NbtAddress getByName( String host ) {
+        return getByName( new Name( host, 0x20, null ));
+    }
+
+    synchronized static NbtAddress getByName( Name name ) {
+        NbtAddress result = null;
+
+        try {
+            if( FILENAME != null ) {
+                File f = new File( FILENAME );
+                long lm;
+
+                if(( lm = f.lastModified() ) > lastModified ) {
+                    lastModified = lm;
+                    TAB.clear();
+                    alt = 0;
+                    populate( new FileReader( f ));
+                }
+                result = (NbtAddress)TAB.get( name );
+            }
+        } catch( FileNotFoundException fnfe ) {
+            if( log.level > 1 ) {
+                log.println( "lmhosts file: " + FILENAME );
+                fnfe.printStackTrace( log );
+            }
+        } catch( IOException ioe ) {
+            if( log.level > 0 )
+                ioe.printStackTrace( log );
+        }
+        return result;
+    }
+
+    static void populate( Reader r ) throws IOException {
+        String line;
+        BufferedReader br = new BufferedReader( r );
+
+        while(( line = br.readLine() ) != null ) {
+            line = line.toUpperCase().trim();
+            if( line.length() == 0 ) {
+                continue;
+            } else if( line.charAt( 0 ) == '#' ) {
+                if( line.startsWith( "#INCLUDE " )) {
+                    line = line.substring( line.indexOf( '\\' ));
+                    String url = "smb1:" + line.replace( '\\', '/' );
+
+                    if( alt > 0 ) {
+                        try {
+                            populate( new InputStreamReader( new SmbFileInputStream( url )));
+                        } catch( IOException ioe ) {
+                            log.println( "lmhosts URL: " + url );
+                            ioe.printStackTrace( log );
+                            continue;
+                        }
+
+                        /* An include was loaded successfully. We can skip
+                         * all other includes up to the #END_ALTERNATE tag.
+                         */
+
+                        alt--;
+                        while(( line = br.readLine() ) != null ) {
+                            line = line.toUpperCase().trim();
+                            if( line.startsWith( "#END_ALTERNATE" )) {
+                                break;
+                            }
+                        }
+                    } else {
+                        populate( new InputStreamReader( new SmbFileInputStream( url )));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the resource created by the following constructor call: *InputStreamReader*. Currently, there are execution paths that do not contain closure statements, for example, when Lmhosts.populate throws an exception. Either a) close the object returned by *InputStreamReader()* in a try-finally block or b) close the resource by declaring the object returned by *InputStreamReader()* in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).